### PR TITLE
`NumberControl`: commit (and constrain) value on `blur` event

### LINKED ISSUE #33291: components: Add baseline validation to NumberControl
## What problem does this address?
It's currently possible to input invalid values into the `NumberControl`: https://github.com/WordPress/gutenberg/pull/33285

To prevent this, you currently have to manage that in the `onChange` you pass. But we should be able to introspect into the `min` and `max` values _at minimum_ to create a generic validation function which prevents invalid values from being entered.

## What is your proposed solution?
Using the `min` and `max` props, verify that the input value is between those numbers, inclusive of both (to match the HTML API).

### PR DESCRIPTION
**Pull request by ciampo**
## Description

Flagged in #33285

Closes #33291

Changes in this PR:

- the behavior of the `InputControl` component when losing focus (`blur` event):
  - previously, the value was committed only if the `isPressEnterToChange` prop was set to `true`
  - after the change from this PR, the value is always committed to the internal state reducer
- `NumberControl`'s internal state reducer constrains the incoming value also when receiving an `UPDATE` action

This change allows the `NumberControl` validation (which is already in place) to kick in when the input loses focus.

## Testing Instructions

- On `NumberControl`:
  - Set `min` and `max` values
  - Input a value that is out of the `min` and `max` bounds
  - Click outside of the input (or press the `tab` key to shift focus away from the input)
  - Notice how the value in the input field updates and gets clamped within the min/max boundaries

- The unit test added in this PR:
  - Check that it passes
  - Undo the changes that this PR makes to `InputControl` and `NumberControl`, and notice how the test fails

- Check, in Storybook, that all usages of `InputControl` and `NumberControl` work as expected (eg `UnitControl`, `BoxControl`, `RangeControl`...)
- Check that all usages of `NumberControl` in Gutenberg work as expected
- Make sure that all unit tests continue to pass

## Screenshots 

Before:

https://user-images.githubusercontent.com/1083581/156559639-29344d95-1f2b-4aa0-bb39-d99b92f66318.mp4

After:

https://user-images.githubusercontent.com/1083581/156559653-8aede884-9c18-4f3b-b7e5-04489fe7a68c.mp4

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- N/A I've updated related schemas if appropriate.

### CODE REVIEW

File: packages/components/src/number-control/index.js
Comment (ciampo) [inline packages/components/src/number-control/index.js]: I'm hesitant about this change.

In the [input control reducer](https://github.com/WordPress/gutenberg/blob/7a3e37a3e0f11e0a05eb2ab36d34427e1c0c5b4f/packages/components/src/input-control/reducer/reducer.ts), the reducers for `COMMIT` and `UPDATE` do the exact same thing, and so I wonder if the `UPDATE` action is necessary at all ?

**Comment by stokesman** [comment]
Thanks for the ping @ciampo. This PR does as advertised but, as I see it, isn't a full fix for the issue. That's because this still calls `onChange` with the values that are out of range. Regarding the original flagging of the issue, the query block would still have to guard against the out of range values in the `onChange` it supplies.

<details><summary>In case you'd agree here's a test for the behavior I'm suggesting:</summary>

```javascript
		it( 'should not call onChange callback with invalid values', () => {
			const spy = jest.fn();
			render(
				<NumberControl
					value={ 5 }
					min={ 2 }
					max={ 10 }
					onChange={ ( v ) => spy( v ) }
				/>
			);
			const input = getInput();
			input.focus();
			fireEvent.change( input, { target: { value: 1 } } );
			expect( input.value ).toBe( '1' );
			expect( spy ).not.toHaveBeenCalled();
		} );
```
</details>

File: packages/components/src/number-control/index.js
Comment (andrewserong) [reply packages/components/src/number-control/index.js]: It does look like that, doesn't it? From the looks of it, we might be able to remove the `update()` action as well, and only use `commit()`, if I'm not missing anything?

**Review by andrewserong — COMMENTED** [review]
This change is testing well for me manually, just left a small comment on the update action.

> the query block would still have to guard against the out of range values in the onChange it supplies.

That's a good question!

I was wondering if there could also be cases where a consuming component should receive `onChange` calls while the user is typing (to display an error message or warning, or update the UI in some way), so we might not want to completely constrain it?

For consumers like the Query block that need to be that careful about values that are used, it could still be worth those consumers using their own validation for the appropriate values. Looking at the change back in https://github.com/WordPress/gutenberg/pull/33285, I'd personally lean toward keeping the guards in the query block's toolbar there as they look like good defensive conditionals to me. But I might be overly wary of relying on component behaviour, so don't feel too strongly about it! 😀

File: packages/components/src/input-control/input-field.tsx
Comment (stokesman) [inline packages/components/src/input-control/input-field.tsx]: Instead of moving the call to `handleOnCommit` out of the conditional it should work better to expand the conditions to enter that block when the input has an invalid value.
```javascript
		if ( isDirty || ! event.target.validity.valid ) {
```
Doing so will make the `wasDirtyOnBlur` ref `true` in the next render and the effect hook will propagate the value instead of taking it from props (where it has not yet been constrained). This obviates the change in the reducer for `NumberControl` (back in 863f33c61cd5a3c2c48b12c7b4b58a5652ac2f12).

Removing `isPressEnterToChange` from the condition is tangential but should be fine because it's a precondition for `isDirty` ever becoming `true` in the first place.

File: packages/components/src/number-control/index.js
Comment (stokesman) [inline packages/components/src/number-control/index.js]: There would be no need to add `UPDATE` here with the other suggestion I left.
UPDATE: I'll mark as resolved since it is outdated.

File: packages/components/src/number-control/index.js
Comment (ciampo) [reply packages/components/src/number-control/index.js]: I went ahead and:
- removed he `UPDATE` action and every reference to it
- swapped `update()` with `commit()` in `InputControl`

(Done in https://github.com/WordPress/gutenberg/pull/39186/commits/e8b3d48cafb32f65a9dfb7b3e072f3869b9c118f)

**Comment by ciampo** [comment]
Thank you @andrewserong and @stokesman !

I should add, as a preface, that I'm not the original author of the `InputControl` and `NumberControl` components, nor I have worked on them much in the past. These components are quite complex with their internal state reducer (to the point that I often wondered if this level of complexity could have been avoided in the first place) and, from what I see from the repo's history, went through several tweaks over the past couple of years. This is to say that I may sometimes miss some nuances, or in general not have a 100% clear picture.

Regarding the firing of `onChange` with un-clamped values, I think we have 3 alternatives:

1. Keep the behaviour as-is — this is the current behaviour, and seems to be the vanilla `input` behaviour as well (see this [codesanbox](https://codesandbox.io/s/boring-tu-nqvlr0?file=/src/App.js)). @andrewserong makes a good point here regarding the fact that consumers of `NumberControl` may rely on the `onChange` callback to read the "raw" value and use it for tasks like validation.
    - Idea: To improve the situation, we could add a new property to [the second argument of `onChange`](https://github.com/WordPress/gutenberg/blob/2df751aa2ad079b9c015bc159720d61b48780e0c/packages/components/src/input-control/types.ts#L35-L38), something like a `isValid` flag, which consumers of `NumberControl` could use instead of checking the value manually
2. Tweak `NumberControl` so that the `onChange` callback doesn't fire if the `value` is out or the min/max range. 
    - Idea: we could add a new prop to `NumberControl` that would enable this new behaviour, so that it becomes an opt-in feature.
3. Tweak `NumberControl` so that the `onChange` callback fires with the clamped `value`
    - Idea: like for the previous alternative, we could add a new prop to `NumberControl` that would enable this new behaviour, so that it becomes an opt-in feature.


Personally I think that the best compromise is option is number 1 (with the addition of `isValid`), but I'm curious to hear your thoughts as well (and @mirka's too!)

**Review by stokesman — COMMENTED** [review]
NOTE: I had most of this review pending before e8b3d48 though it applies well enough still.

I took a closer look at this and suspected some unintended behavior. I added a `console.log` to the Storybook example and here's what I found:

Mar-07-2022 23-14-03

The `onChange` with the constrained value does not fire until the input gains focus again. I've left a few notes on how to fix it up.

Still, I'm not convinced there is much value in this because constraint of an invalid value on `blur` is already a given in (typical) controlled contexts. That's because `NumberControl` intakes the value from props once it's no longer focused (given that `valueProp !== value`). It looks useful in the Storybook example because, while technically controlled, the wrapping component doesn't limit its own value.

**Comment by stokesman** [comment]
Great to see the consideration of the options @ciampo.

Regarding the 1st option and an `isValid` flag, I'd like to note that it already exists in the more verbose `event.target.validity.valid` (because the event is part of that second argument). Adding `isValid` could be convenient though.

The 2nd and 3rd options could also be combined. That is, the `onChange` could cease fire while the input holds an invalid value and fire with the resolved value on `blur`. That is what I've tried in my most recent take on this in #39260.

With regards to the excellent point raised by @andrewserong about the need of consuming components to respond to any input. It happens that `RangeControl` is one such example. When out-of-range numbers are entered in its `NumberControl` it updates its slider to the min/max limit (and propagates the new value). So if `NumberControl` were to stop calling `onChange` with invalid values that would break. An alternative way for consumers to respond to such values is through expanded use of the existing `onValidate` prop (as in #39260). Not that I am convinced it is the best way.

**Comment by mirka** [comment]
Option 1 sounds good to me too! No strong opinion on whether to add a `isValid` for convenience or use the existing `event.target.validity.valid`, as long as we clearly document the pattern.

**Comment by andrewserong** [comment]
Thanks for the detailed follow-up @ciampo! These components are particularly tricky, and I know what you mean, it's hard getting into some of the details when we weren't the folks to originally write the code — I'd love it if longer-term we could reduce some of the complexity, too. Thanks for digging in!

> Option 1 sounds good to me too! No strong opinion on whether to add a isValid for convenience or use the existing event.target.validity.valid, as long as we clearly document the pattern.

Same, I like option 1, and the addition of an `isValid` flag in the callback sounds like a good compromise to me — it still gives consumers a clear way to rely on the component for validation if they want to defer validation, as well as the raw value if they want to handle validation within the consuming code.

File: packages/components/src/number-control/index.js
Comment (andrewserong) [reply packages/components/src/number-control/index.js]: Thanks, those changes look good to me!

File: packages/components/src/input-control/input-field.tsx
Comment (andrewserong) [reply packages/components/src/input-control/input-field.tsx]: Oh, good catch about the case of `onChange` being called with the constrained value only when the input gains focus again!

Comment (ciampo) [reply packages/components/src/input-control/input-field.tsx]: Good point, @stokesman — makes a lot of sense!

I've applied your suggestion (https://github.com/WordPress/gutenberg/pull/39186/commits/cbf0beffb1d3d6d77d7d0382db7e44b1c5d06c80) and I've also added a unit test to check the behaviour and prevent future regressions (https://github.com/WordPress/gutenberg/pull/39186/commits/80982d0cb9d69d40cae4fa67bba80283ed298e18)

**Comment by ciampo** [comment]
Thank you all for the last round of feedback — I should have fixed the issue with `onChange`, the callback should now fire also when the value gets clamped on `blur`.

Regarding the behavior of the `onChange`, for now I've adopted what we called "option 1", but _without_ adding a new `isValid` prop — instead, I've updated docs, Storybook examples and unit tests to clearly reference the `event.target.validity.valid` value (https://github.com/WordPress/gutenberg/pull/39186/commits/d91e6a46175b1e4d6908e40ed6f15967b3214c08). I think that relying on native browser APIs here may be better on the long term (less custom code to maintain, less conventions).

One thing that we may want to consider, finally, is whether consumers of `NumberControl` (like `RangeControl`, `UnitControl`...) should also expose a similar "validity" flag.

Let me know what you all think!

File: packages/components/src/number-control/README.md
Comment (andrewserong) [inline packages/components/src/number-control/README.md:107]: I think I have a slight (though not at all strong) preference for having the second param be the plain event object rather than nesting it within an `extra` object. There's a familiarity with having the second param of callbacks be the event object itself, so it might make this component slightly easier for someone new to looking at it if we don't add in this abstraction.

But I'm aware you probably added it in mostly thinking of the `isValid` flag? I think given the documentation now covers `event.target.validity.valid`, we'd probably be fine not adding in any additional metadata.

What do you think?

File: packages/components/src/number-control/stories/index.js
Comment (andrewserong) [inline packages/components/src/number-control/stories/index.js:52]: Nice way of visualising this in Storybook 👍

**Review by andrewserong — APPROVED** [review]
This is testing well for me @ciampo, thanks for updating the Storybook example, too! (Tested a few UnitControl components within the editor, and the Storybook examples). ✨

> I think that relying on native browser APIs here may be better on the long term (less custom code to maintain, less conventions).

That makes good sense to me!

Just left an optional comment about using the `event` object directly as the second param in the callback instead of wrapping in an `extra` object (so folks don't have to remember to destructure it), but I don't feel strongly about it, happy to go with whatever everyone thinks makes most sense here!

File: packages/components/src/number-control/README.md
Comment (stokesman) [reply packages/components/src/number-control/README.md:107]: Chiming in to say the shape of that goes back well before this PR. I've suggested the same https://github.com/WordPress/gutenberg/pull/33696#discussion_r677835665. My feeling is the components are experimental so we have license to make the change as long as we fix up any internal stuff that might break. Good for a separate PR though.

File: packages/components/src/input-control/input-field.tsx
Comment (stokesman) [reply packages/components/src/input-control/input-field.tsx]: Nice tests!

Something we might want to think about with regards to `event.target.validity.valid` is that it will be `false` for `step` mismatches and [those can occur even without typing them](34566) but I don't really know if it's a concern here or not.

Comment (ciampo) [reply packages/components/src/input-control/input-field.tsx]: That's a good point!

I also think it's not really a concern here — I'd personally keep it as-is and only make adjustments if needed (plus, #34566 could also help if/when it gets merged). I still [made a little change to the README](https://github.com/WordPress/gutenberg/pull/39186/commits/1285c69b1ae824cea7c328d7d9629d8a01915d84) to also illustrate the `step` mismatch scenario that you pointed out.

After all, if `event.target.validity.valid` is too strict, folks can also implement their own "validation" (which is that's happening at the moment)

File: packages/components/src/number-control/README.md
Comment (ciampo) [reply packages/components/src/number-control/README.md:107]: As @stokesman mentioned, the `extra` object existed before this PR and [is defined in `InputControl`](https://github.com/WordPress/gutenberg/blob/2df751aa2ad079b9c015bc159720d61b48780e0c/packages/components/src/input-control/types.ts#L35-L38).

For example, `UnitControl` [extends that `extra` object ](https://github.com/WordPress/gutenberg/blob/0ea49d674a376636db5c0722ae41083394df9e60/packages/components/src/unit-control/types.ts#L40-L43)by adding a custom `data` property too (which seems to include extra info about which unit is being selected).

Given the above, I'm not sure what would be the best way to simplify the signature of the `onChange` callback — but it's definitely something that can be done in a follow-up

**Comment by ciampo** [comment]
Just rebased on top of `trunk`. I'm going to wait until tomorrow for tests to pass and any further feedback to come through, and otherwise I'll merge

File: packages/components/src/number-control/README.md
Comment (andrewserong) [reply packages/components/src/number-control/README.md:107]: Ah, excellent, thanks for linking the prior discussion, glad I'm not alone there! 😄

Yes, I think that's perfectly fine for a follow-up, not a blocker for me!

**Comment by ciampo** [comment]
As I was giving a final look at the PR after rebasing, I found an issue with the current approach:

Basically, when dragging gestures, the `event` emitted via the `onChange` callback by `InputControl` is a `PointerEvent` (contrary to what we [typed](https://github.com/WordPress/gutenberg/blob/2df751aa2ad079b9c015bc159720d61b48780e0c/packages/components/src/input-control/types.ts#L36)) that doesn't always have the `input` HTML element as its target — in fact, in my tests it can also be its parent `div`, the `body` or even the `html` element (depending on where is the pointer on the page).

https://user-images.githubusercontent.com/1083581/158367477-ac036ff6-06e0-4789-a454-c442eaf1fd20.mp4

This causes 2 issues:

- our types for `InputChangeCallback` are wrong:
    - technically the event can be any type of event (`KeyboardEvent`, `PointerEvent`...)
    - the event target is not guaranteed to be of type `HTMLInputEvent`
- As a consequence, the `event.target.validity.valid` approach won't work when the event target is not the `input` element (to the point that the current Storybook example for `NumberControl` will stop working as soon as the pointer is dragged beyond the story boundaries)

I'm not sure if we can (or want to) change the behaviour of `useDrag` to only fire events when the `input` element is the target.

We should probably update our types (and docs) to reflect the correct types for the `onChange` callbacks.

Regarding the `valid` flag, I'm not really sure about what's the best approach here:

- the simpler (but less elegant) solution is to tweak docs and Storybook examples to check first if the `validity` key is defined on `event.target`. We could add that, in case of a pointer event, the value can be assumed valid (since the value is [already clamped/validated](https://github.com/WordPress/gutenberg/blob/021979306cddf4acfab901b3f33e2b15c05ab486/packages/components/src/number-control/index.js#L143-L146) in the `NumberControl` reducer)
- alternatively, we could reconsider adding a `isValid` convenience flag to `NumberControl`'s `onChange` callback, and manually check the value in there. This is probably the better solution between the two

Thoughts?

File: packages/components/CHANGELOG.md
Comment (ciampo) [inline packages/components/CHANGELOG.md]: This needs to be moved to the new `Unreleased` section

**Comment by mirka** [comment]
> Regarding the `valid` flag, I'm not really sure about what's the best approach here

I'm fine with taking the first inelegant approach here to land this PR, and then triage the `isValid` approach separately as an enhancement (if at all).

**Comment by ciampo** [comment]
> > Regarding the `valid` flag, I'm not really sure about what's the best approach here
> 
> I'm fine with taking the first inelegant approach here to land this PR, and then triage the `isValid` approach separately as an enhancement (if at all).

Makes sense. I've pushed changes to [update the `onChange` event type](https://github.com/WordPress/gutenberg/pull/39186/commits/196f3b4afc1986ff0ecae1787c708e615284fbc1) and to [update docs and Storybook example](https://github.com/WordPress/gutenberg/pull/39186/commits/fc9fcfd7a7a4fa32c990cf04c6e571cf36cbcb8c) to show that `validity` is not always going to be defined on `event.target`.

**Comment by stokesman** [comment]
Great catch there @ciampo! It got to thinking "what about `currentTarget` on those events?" and tried logging the event to see. I happened to log it from within the `useDrag` handler and found the `currentTarget` and even `target` were always the input! This is because what I was logging was the [gesture state](https://use-gesture.netlify.app/docs/state#gesture-state-attributes) which appears to be an extended event type.

So a possible way to keep the expectation that `validity.value` will always exist is to let the gesture state forward through as the event. I suppose we may prefer to look at that separately though.

**Comment by ciampo** [comment]
> tried logging the event to see. I happened to log it from within the `useDrag` handler and found the `currentTarget` and even `target` were always the input! This is because what I was logging was the [gesture state](https://use-gesture.netlify.app/docs/state#gesture-state-attributes) which appears to be an extended event type.

Nice catch!

Since there seem to be some flaky e2e tests, I decided in the meantime to[ give this approach a try by "overriding" the `event.target` with the `target` from the gesture state](https://github.com/WordPress/gutenberg/pull/39186/commits/ea8f2d5cfa279f62894e4ffb97ebd73ffbcd43ca). From a quick test it seems to work well! Let me know what your thoughts are, happy to roll back to the previous approach if this feels too hacky.

**Comment by stokesman** [comment]
Well, I didn't mean to push to this branch 🤦 but now that it happened maybe it's just as well to leave it for the moment. I meant to put it on a separate branch and link a compare page to show the way I had thought to approach the latest suggestion I'd made. I'm not sure there's anything strictly better about it but thought it could be worth a look. Now front and center! But I'd be happy to drop it from this branch and won't mind if you do.

File: packages/components/src/unit-control/index.tsx
Comment (stokesman) [inline packages/components/src/unit-control/index.tsx]: I didn't intentionally change this. I was working off a no-commit revert and hadn't noticed until now.

File: packages/components/src/input-control/reducer/reducer.ts
Comment (stokesman) [inline packages/components/src/input-control/reducer/reducer.ts]: This is the core difference. When the action one of the drag ones, the full gesture state becomes `_event` on the new state. It looks enough like an event I think and comes with a type we can use 🤷.

**Comment by ciampo** [comment]
> Well, I didn't mean to push to this branch 🤦
> ...
> I'm not sure there's anything strictly better about it but thought it could be worth a look

This approach is definitely interesting and worth considering — on the one hand, it feels "cleaner" and more elegant, but on the other hand it raises a few potential issues:

- in case of a drag gesture, the shape of the event passed to the `onChange` callback would be different from the standard browser event, which could throw a few folks off
- on top of that, we'd be exposing parts of a third-party library (`@use-gesture`'s gesture state) through the public API of `UnitControl`, which is something we usually don't do, as we want to have the freedom to easily swap that library for another solution without causing a breaking change

For these reasons, I'm inclined to rollback to my last approach ("overriding" only the `event.target` with the `target` from the gesture state, leaving the rest of the event object untouched).

**Comment by stokesman** [comment]
>on top of that, we'd be exposing parts of a third-party library (@use-gesture's gesture state) through the public API of UnitControl, which is something we usually don't do, as we want to have the freedom to easily swap that library for another solution without causing a breaking change

I have to agree it'd be better to expose only what we decide to.

>I'm inclined to rollback to my last approach…

Sounds good to me. I'd say feel free to drop my commit right out of history. I would have done so but didn't want to force push on this branch.

**Comment by ciampo** [comment]
Thank you, @stokesman! I dropped your commit as agreed.

I believe this PR is ready for a final check

**Review by andrewserong — APPROVED** [review]
Re-tested, this is looking good to me @ciampo! Min/max constrained values are working correctly, both in Storybook examples, and in the editor (specifically looked at the Spacer height control since we were already looking at that separately). And the tests are passing 👍

LGTM!

**Pull request merged**
